api/krusty/component_test.go

```diff
@@ -0,0 +1,441 @@
+// Copyright 2019 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package krusty_test
+
+import (
+	"strings"
+	"testing"
+
+	kusttest_test "sigs.k8s.io/kustomize/api/testutils/kusttest"
+)
+
+// Test kustomization.yaml files with  kind: Component
+func writeComponentBase(th kusttest_test.Harness) {
+	th.WriteK("/app/base", `
+resources:
+- deploy.yaml
+configMapGenerator:
+- name: my-configmap
+  literals:	
+    - testValue=1
+    - otherValue=10
+`)
+	th.WriteF("/app/base/deploy.yaml", `
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: storefront
+spec:
+  replicas: 1
+`)
+}
+
+func writeComponentPatch(th kusttest_test.Harness) {
+	th.WriteF("/app/patch/kustomization.yaml", `
+apiVersion: kustomize.config.k8s.io/v1alpha1
+kind: Component
+namePrefix: patched-
+replicas:
+- name: storefront
+  count: 3
+resources:
+  - stub.yaml
+configMapGenerator:
+- name: my-configmap
+  behavior: merge
+  literals:	
+    - testValue=2
+    - patchValue=5
+`)
+	th.WriteF("/app/patch/stub.yaml", `
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: stub
+spec:
+  replicas: 1
+`)
+}
+
+func writeComponentProd(th kusttest_test.Harness) {
+	th.WriteK("/app/prod", `
+resources:
+- ../base
+- db
+
+components:
+- ../patch
+`)
+	th.WriteF("/app/prod/db", `
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: db
+spec:
+  type: Logical
+`)
+}
+
+// Components are inserted into the resource hierarchy as the parent of those
+// resources that come before it in the resources list of the parent Kustomization.
+func TestBasicComponent(t *testing.T) {
```

Comment:
Thanks for these tests!  Would you consider making them table-driven tests?  I know a lot of the existing tests aren't table driven, but a lot of the new tests are being written this way.

https://dave.cheney.net/2019/05/07/prefer-table-driven-tests

Example: https://github.com/kubernetes-sigs/kustomize/blob/master/api/filters/annotations/annotations_test.go#L19

Reply:
OK, will do - sounds like a good idea.

Reply:
Done - thanks for the link - do they look ok?  Could maybe think of moving some of the helper classes to `harness.go`, and maybe even moving the error-generating tests to another file?